### PR TITLE
[WIP] Add some support for OpenTracing API

### PR DESCRIPTION
**Pull request by Quid37**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

OpenTracing API support

Detailed description / Documentation draft:

...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

**Review by qoega — COMMENTED** [review]
Have you checked if it is hard to switch from OpenTracing-> OpenTelemetry or they are compatible?
I believe that OpenTracing will be supported further, but probably it is better to use their newer project.

File: programs/server/HTTPHandler.cpp [hunk 1]
```diff
@@ -540,6 +540,28 @@ void HTTPHandler::processQuery(
...
+        if (!client_info.parent_span_context_string.empty()) {
+            // TODO: support other SpanReferenceType.
+            auto parent_span_context = std::make_shared<opentracing::SpanContext>(context.getDistributedTracer()->DeserializeSpanContext(client_info.parent_span_context_string));
+            references.push_back({opentracing::SpanReferenceType::ChildOfRef, std::move(parent_span_context)});
+        }
+
+        /// If we are in initial query then we may initialize new trace with some probability.
+        if (!references.empty()
+            || (client_info.query_kind == ClientInfo::QueryKind::INITIAL_QUERY
+                && std::uniform_real_distribution<>()(thread_local_rng) <= settings.distributed_tracing_probability))
```
[the remark this answers is not in the document]
Comment (alexey-milovidov) [reply programs/server/HTTPHandler.cpp:558]: > std::uniform_real_distribution<>()(thread_local_rng) <= settings.distributed_tracing_probability

bernoulli_distribution.

File: src/Common/DistributedTracing.h [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+    virtual SpanContext DeserializeSpanContext(const String& span_context_string) const = 0;
+};
+
+
+class DistributedTracer
+    : public IDistributedTracer
+{
+public:
+    // TODO add tags to span class.
```
[the remark this answers is not in the document]
Comment (Quid37) [reply src/Common/DistributedTracing.h:146]: Yes

File: src/Core/Settings.h [hunk 3]
```diff
@@ -154,6 +154,10 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingLogQueriesType, log_queries_min_type, QueryLogElementType::QUERY_START, "query_log minimal type to log, possible values (from low to high): QUERY_START, QUERY_FINISH, EXCEPTION_BEFORE_START, EXCEPTION_WHILE_PROCESSING.", 0) \
     M(SettingUInt64, log_queries_cut_to_length, 100000, "If query length is greater than specified threshold (in bytes), then cut query when writing to query log. Also limit length of printed query in ordinary text log.", 0) \
     \
+    M(SettingBool, enable_distributed_tracing, false, "Produce OpenTracing-compatible tracing (implemented over query_log)", 0) \
+    M(SettingBool, use_noop_tracer, false, "Use special NoopTracer (may be useful for tests or to avoid tracing specific parts of code)", 0) \
+    M(SettingFloat, distributed_tracing_probability, 1., "Probability of a request to be traced (in case distributed tracing is enabled)", 0) \
```
[the remark this answers is not in the document]
Comment (Quid37) [reply src/Core/Settings.h:159]: That's true. I just put it to 1 to ease testing, but will change it when this pull request will stop being draft.

**Comment by Quid37** [comment]
> Have you checked if it is hard to switch from OpenTracing-> OpenTelemetry or they are compatible?
> I believe that OpenTracing will be supported further, but probably it is better to use their newer project.

According to the roadmap https://medium.com/opentracing/a-roadmap-to-convergence-b074e5815289, they will support OpenTracing compatibility for 2 more years after the release. It is desirable to migrate, but I have no idea when C++ library will be realeased: https://github.com/open-telemetry/opentelemetry-cpp/blob/master/README.md

**Comment by akuzm** [comment]
@Quid37 do you plan to continue working on this?

I think distributed tracing may be useful to debug problems in large cluster, related to communication between different server. This means we should pay most of attention to the events related to inter-server communication. This includes queries made during distributed group by, remote functions such as `S3`, foreign dictionaries, etc. On the other hand, duplicating the details about the specific stages of query execution seems less useful -- we already have query log, query thread log, `EXPLAIN`, query profiler etc. to evaluate the performance of the specific stages of the query. Maybe a general overview of SQL stages like `SELECT/WHERE/GROUP BY/HAVING/ORDER BY` can be helpful, but not more verbose than that.

So the way forward I see for this is to make an MVP that publishes spans for the entire query + spans for network requests to other ClickHouse servers in cluster, and merge it. Then we can experiment with it, add other spans such as distributed functions and dictionaries, etc.

Another interesting direction is to see whether we can save all this in some ClickHouse table. Currently, to debug the distributed queries, we scan system logs on cluster with queries like `select ... from remote('cluster', system, query_log)`, which has certain drawbacks such that it doesn't reach all shards or some interesting shards might be down. Having a special log table on query coordinator that saves the span info would be convenient.

**Comment by akuzm** [comment]
We added some OpenTelemetry support based on this prototype, so this PR can be closed now.

https://github.com/ClickHouse/ClickHouse/pull/14195
https://github.com/ClickHouse/ClickHouse/pull/16535

**Pull request closed without merging**